File: ios/Classes/datacapture/SupyDataCaptureScannerView.swift
import AVFoundation
import Flutter
import UIKit
import Vision

/// PlatformView hosting an `AVCaptureSession` preview for the embedded live
/// text-pattern (generic data-capture) scanner (DC7).
///
/// The native side runs `VNRecognizeTextRequest` per frame and ships the
/// recognized-text geometry (`frame_text` events); the Dart side
/// (`SupyTextPatternMatcher`) runs the regex patterns over it. This view
/// therefore exposes only camera lifecycle: torch / pause / resume + a
/// `preview_started` event — there is no native pattern matching.
///
/// Threading rules (per CLAUDE.md):
///   - `AVCaptureSession` configuration and start/stop run on a dedicated
///     background queue (never `.main`).
///   - `VNRequest` work runs on the detector's background queue.
///   - `FlutterEventSink` invocations are marshalled to `.main`.
final class SupyDataCaptureScannerView: NSObject, FlutterPlatformView,
  FlutterStreamHandler
{

  private let container: PreviewContainerView
  private let session: AVCaptureSession = AVCaptureSession()
  private let sessionQueue: DispatchQueue = DispatchQueue(
    label: "io.supy.scanner.datacapture.session",
    qos: .userInitiated
  )
  private var previewLayer: AVCaptureVideoPreviewLayer?
  private var videoDevice: AVCaptureDevice?

  private let methodChannel: FlutterMethodChannel
  private let eventChannel: FlutterEventChannel

  private var eventSink: FlutterEventSink?
  private var previewStartedAnnounced: Bool = false
  private var sessionConfigured: Bool = false
  private var runningObservation: NSKeyValueObservation?
  private var sessionInterrupted: Bool = false

  private let detector: TextFrameDetector
  private let videoDataOutput: AVCaptureVideoDataOutput = AVCaptureVideoDataOutput()

  init(
    frame: CGRect,
    viewId: Int64,
    creationParams: [String: Any?]?,
    messenger: FlutterBinaryMessenger
  ) {
    self.container = PreviewContainerView(frame: frame)
    self.container.backgroundColor = .black

    let prefix = "io.supy.scanner/v1/datacapture"
    self.methodChannel = FlutterMethodChannel(
      name: "\(prefix)/\(viewId)",
      binaryMessenger: messenger
    )
    self.eventChannel = FlutterEventChannel(
      name: "\(prefix)/\(viewId)/events",
      binaryMessenger: messenger
    )

    let languages = (creationParams?["languages"] as? [String]) ?? []
    self.detector = TextFrameDetector(languages: languages)

    super.init()

    self.methodChannel.setMethodCallHandler { [weak self] call, result in
      self?.handle(call: call, result: result)
    }
    self.eventChannel.setStreamHandler(self)

    // Detector callbacks fire on its own queue — marshal to main in sendEvent.
    detector.onFrameText = { [weak self] tree in
      self?.emitFrameText(tree)
    }
    detector.onError = { [weak self] message in
      self?.emitError(code: "unknown", message: message)
    }

    registerSessionNotifications()
    configureSessionAsync()
  }

  private func registerSessionNotifications() {
    let center = NotificationCenter.default
    center.addObserver(
      self,
      selector: #selector(handleSessionWasInterrupted(_:)),
      name: .AVCaptureSessionWasInterrupted,
      object: session
    )
    center.addObserver(
      self,
      selector: #selector(handleSessionInterruptionEnded(_:)),
      name: .AVCaptureSessionInterruptionEnded,
      object: session
    )
    center.addObserver(
      self,
      selector: #selector(handleSessionRuntimeError(_:)),
      name: .AVCaptureSessionRuntimeError,
      object: session
    )
  }

  @objc private func handleSessionWasInterrupted(_ note: Notification) {
    sessionInterrupted = true
  }

  @objc private func handleSessionInterruptionEnded(_ note: Notification) {
    sessionInterrupted = false
    sessionQueue.async { [weak self] in
      self?.safeStartRunning()
    }
  }

  @objc private func handleSessionRuntimeError(_ note: Notification) {
    let error = note.userInfo?[AVCaptureSessionErrorKey] as? NSError
    if isNotAuthorizedError(error) {
      emitError(
        code: "permission_denied",
        message: "Camera permission revoked mid-session."
      )
      return
    }
    emitError(
      code: "camera_unavailable",
      message: "AVCaptureSession runtime error: \(error?.localizedDescription ?? "unknown")"
    )
  }

  private func isNotAuthorizedError(_ error: NSError?) -> Bool {
    guard let error = error else { return false }
    return error.domain == AVFoundationErrorDomain &&
      error.code == AVError.Code.applicationIsNotAuthorizedToUseDevice.rawValue
  }

  /// Always call instead of `session.startRunning()` directly. Must be invoked
  /// from `sessionQueue`.
  private func safeStartRunning() {
    guard !sessionInterrupted else { return }
    guard !session.isRunning else { return }
    session.startRunning()
  }

  // MARK: - FlutterPlatformView

  func view() -> UIView {
    return container
  }

  // MARK: - Session lifecycle

  private func configureSessionAsync() {
    let status = AVCaptureDevice.authorizationStatus(for: .video)
    switch status {
    case .authorized:
      sessionQueue.async { [weak self] in self?.configureSession() }
    case .notDetermined:
      AVCaptureDevice.requestAccess(for: .video) { [weak self] granted in
        guard let self = self else { return }
        if granted {
          self.sessionQueue.async { [weak self] in self?.configureSession() }
        } else {
          self.emitError(
            code: "permission_denied",
            message: "Camera permission denied by the user."
          )
        }
      }
    case .denied, .restricted:
      emitError(
        code: "permission_denied",
        message: "Camera authorization not granted (status: \(status.rawValue))."
      )
    @unknown default:
      emitError(
        code: "camera_unavailable",
        message: "Unknown camera authorization status (\(status.rawValue))."
      )
    }
  }

  private func configureSession() {
    guard !sessionConfigured else { return }

    session.beginConfiguration()
    let tier = SupyDeviceTier.detect()
    session.sessionPreset = SupyDataCaptureScannerView.sessionPreset(for: tier)

    guard
      let device = AVCaptureDevice.default(
        .builtInWideAngleCamera, for: .video, position: .back
      )
    else {
      session.commitConfiguration()
      emitError(code: "camera_unavailable", message: "No back camera available.")
      return
    }
    videoDevice = device

    do {
      let input = try AVCaptureDeviceInput(device: device)
      if session.canAddInput(input) {
        session.addInput(input)
      } else {
        session.commitConfiguration()
        emitError(
          code: "camera_unavailable",
          message: "Cannot attach camera input to capture session."
        )
        return
      }
    } catch {
      session.commitConfiguration()
      emitError(
        code: "camera_unavailable",
        message: "Camera input init failed: \(error.localizedDescription)"
      )
      return
    }

    videoDataOutput.alwaysDiscardsLateVideoFrames = true
    videoDataOutput.videoSettings = [
      kCVPixelBufferPixelFormatTypeKey as String:
        kCVPixelFormatType_420YpCbCr8BiPlanarFullRange
    ]
    videoDataOutput.setSampleBufferDelegate(
      detector,
      queue: detector.sampleBufferQueue
    )
    if session.canAddOutput(videoDataOutput) {
      session.addOutput(videoDataOutput)
    }

    SupyDataCaptureScannerView.applyFpsCap(tier: tier, device: device)

    session.commitConfiguration()
    sessionConfigured = true

    DispatchQueue.main.async { [weak self] in
      self?.attachPreviewLayer()
    }

    runningObservation = session.observe(\.isRunning, options: [.new]) {
      [weak self] _, change in
      guard let self = self, let running = change.newValue, running else { return }
      if !self.previewStartedAnnounced {
        self.previewStartedAnnounced = true
        self.emitPreviewStarted()
      }
    }

    safeStartRunning()
  }

  private func attachPreviewLayer() {
    let layer = AVCaptureVideoPreviewLayer(session: session)
    layer.videoGravity = .resizeAspectFill
    container.attach(previewLayer: layer)
    previewLayer = layer
  }

  // MARK: - Events

  private func emitPreviewStarted() {
    let flashAvailable = videoDevice?.hasTorch == true
    sendEvent([
      "type": "preview_started",
      "flashAvailable": flashAvailable,
    ])
  }

  private func emitFrameText(_ tree: [String: Any]) {
    var payload = tree
    payload["type"] = "frame_text"
    sendEvent(payload)
  }

  private func emitError(code: String, message: String) {
    sendEvent([
      "type": "error",
      "code": code,
      "message": message,
    ])
  }

  private func sendEvent(_ payload: [String: Any?]) {
    let dispatch: () -> Void = { [weak self] in
      guard let sink = self?.eventSink else { return }
      sink(payload)
    }
    if Thread.isMainThread {
      dispatch()
    } else {
      DispatchQueue.main.async(execute: dispatch)
    }
  }

  // MARK: - MethodChannel

  private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
    dispatchPrecondition(condition: .onQueue(.main))
    switch call.method {
    case "pause":
      sessionQueue.async { [weak self] in
        self?.session.stopRunning()
        DispatchQueue.main.async { result(nil) }
      }
    case "resume":
      sessionQueue.async { [weak self] in
        self?.safeStartRunning()
        DispatchQueue.main.async { result(nil) }
      }
    case "setTorch":
      let on = (call.arguments as? [String: Any])?["on"] as? Bool ?? false
      setTorch(on: on, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func setTorch(on: Bool, result: @escaping FlutterResult) {
    sessionQueue.async { [weak self] in
      guard let device = self?.videoDevice, device.hasTorch else {
        DispatchQueue.main.async { result(nil) }
        return
      }
      do {
        try device.lockForConfiguration()
        device.torchMode = on ? .on : .off
        device.unlockForConfiguration()
      } catch {
        // Swallow — parity with the barcode view.
      }
      DispatchQueue.main.async { result(nil) }
    }
  }

  // MARK: - FlutterStreamHandler

  func onListen(
    withArguments arguments: Any?,
    eventSink events: @escaping FlutterEventSink
  ) -> FlutterError? {
    self.eventSink = events
    if session.isRunning && !previewStartedAnnounced {
      previewStartedAnnounced = true
      emitPreviewStarted()
    }
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    self.eventSink = nil
    return nil
  }

  // MARK: - Teardown

  deinit {
    methodChannel.setMethodCallHandler(nil)
    if let sink = eventSink {
      eventSink = nil
      if Thread.isMainThread {
        sink(FlutterEndOfEventStream)
      } else {
        DispatchQueue.main.async { sink(FlutterEndOfEventStream) }
      }
    }
    eventChannel.setStreamHandler(nil)
    runningObservation?.invalidate()
    runningObservation = nil
    NotificationCenter.default.removeObserver(self)
    videoDataOutput.setSampleBufferDelegate(nil, queue: nil)
    let session = self.session
    sessionQueue.async {
      if session.isRunning {
        session.stopRunning()
      }
    }
  }

  // MARK: - Perf tier

  private static func sessionPreset(for tier: SupyDeviceTier) -> AVCaptureSession.Preset {
    switch tier {
    case .high: return .high
    case .mid: return .hd1280x720
    case .low: return .vga640x480
    }
  }

  private static func applyFpsCap(tier: SupyDeviceTier, device: AVCaptureDevice) {
    // OCR is heavier than barcode decode, so always cap frame delivery — use
    // the per-tier cap where present, otherwise a conservative 10fps ceiling.
    let fps = tier.analyzerFpsCap ?? 10
    do {
      try device.lockForConfiguration()
      let duration = CMTime(value: 1, timescale: CMTimeScale(fps))
      device.activeVideoMinFrameDuration = duration
      device.activeVideoMaxFrameDuration = duration
      device.unlockForConfiguration()
    } catch {
      // Best-effort; perf cap is advisory.
    }
  }
}

/// Per-frame Vision text recognizer feeding the data-capture view.
///
/// Runs `VNRecognizeTextRequest` synchronously on `sampleBufferQueue` for each
/// delivered frame and builds the block → line → element tree consumed by
/// `SupyRecognizedText.fromMap`. Mirrors `OcrRunner.recognizeStructured`:
/// Vision has no block concept, so all lines wrap into one synthetic block
/// whose box is the union of its lines; boxes flip from Vision's bottom-left
/// to Supy's top-left `[0..1]` convention.
final class TextFrameDetector: NSObject, AVCaptureVideoDataOutputSampleBufferDelegate {

  let sampleBufferQueue = DispatchQueue(
    label: "io.supy.scanner.datacapture.detect",
    qos: .userInitiated
  )

  var onFrameText: (([String: Any]) -> Void)?
  var onError: ((String) -> Void)?

  private let languages: [String]

  init(languages: [String]) {
    self.languages = languages
    super.init()
  }

  func captureOutput(
    _ output: AVCaptureOutput,
    didOutput sampleBuffer: CMSampleBuffer,
    from connection: AVCaptureConnection
  ) {
    guard let pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else { return }

    let request = VNRecognizeTextRequest()
    // Live path: prioritize latency over the last few percent of accuracy.
    request.recognitionLevel = .fast
    request.usesLanguageCorrection = false
    if !languages.isEmpty {
      request.recognitionLanguages = languages
    }

    let handler = VNImageRequestHandler(
      cvPixelBuffer: pixelBuffer,
      // Back camera in portrait delivers landscape-right buffers; `.right`
      // presents upright text to Vision.
      orientation: .right,
      options: [:]
    )
    do {
      try handler.perform([request])
    } catch {
      onError?("Text recognition failed: \(error.localizedDescription)")
      return
    }

    let tree = Self.buildTree(
      observations: request.results ?? [],
      includeElements: true
    )
    onFrameText?(tree)
  }

  private static func buildTree(
    observations: [VNRecognizedTextObservation],
    includeElements: Bool
  ) -> [String: Any] {
    guard !observations.isEmpty else {
      return ["fullText": "", "blocks": []]
    }

    var lines: [[String: Any]] = []
    var lineTexts: [String] = []
    var union: CGRect?

    for obs in observations {
      guard let candidate = obs.topCandidates(1).first else { continue }
      let text = candidate.string
      lineTexts.append(text)
      union = union.map { $0.union(obs.boundingBox) } ?? obs.boundingBox

      var elements: [[String: Any]] = []
      if includeElements {
        elements = words(in: candidate, text: text)
      }
      lines.append([
        "text": text,
        "boundingBox": normRect(obs.boundingBox),
        "elements": elements,
      ])
    }

    let fullText = lineTexts.joined(separator: "\n")
    let block: [String: Any] = [
      "text": fullText,
      "boundingBox": normRect(union ?? .zero),
      "lines": lines,
    ]
    return ["fullText": fullText, "blocks": [block]]
  }

  private static func words(
    in candidate: VNRecognizedText,
    text: String
  ) -> [[String: Any]] {
    var out: [[String: Any]] = []
    var idx = text.startIndex
    while idx < text.endIndex {
      while idx < text.endIndex, text[idx].isWhitespace {
        idx = text.index(after: idx)
      }
      guard idx < text.endIndex else { break }
      let start = idx
      while idx < text.endIndex, !text[idx].isWhitespace {
        idx = text.index(after: idx)
      }
      let word = String(text[start..<idx])
      let box: CGRect
      if let rect = try? candidate.boundingBox(for: start..<idx)?.boundingBox {
        box = rect
      } else {
        box = .zero
      }
      out.append(["text": word, "boundingBox": normRect(box)])
    }
    return out
  }

  private static func normRect(_ bb: CGRect) -> [String: Double] {
    return [
      "left": Double(bb.origin.x),
      "top": Double(1.0 - bb.origin.y - bb.size.height),
      "width": Double(bb.size.width),
      "height": Double(bb.size.height),
    ]
  }
}

File: android/src/main/kotlin/io/supy/scanner/datacapture/SupyDataCaptureScannerView.kt
package io.supy.scanner.datacapture

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Color
import android.graphics.Rect
import android.os.Handler
import android.os.Looper
import android.view.View
import android.widget.FrameLayout
import androidx.annotation.MainThread
import androidx.annotation.WorkerThread
import androidx.camera.core.ExperimentalGetImage
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.ImageProxy
import androidx.camera.view.LifecycleCameraController
import androidx.camera.view.PreviewView
import androidx.lifecycle.LifecycleOwner
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.Text
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.TextRecognizer
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView
import io.supy.scanner.barcode.ActivityHolder
import io.supy.scanner.perf.DeviceTier
import java.util.concurrent.Executors

/**
 * PlatformView hosting the embedded live text-pattern (generic data-capture)
 * camera preview + per-frame ML Kit text recognizer (DC7).
 *
 * The native side ships recognized-text geometry per frame (`frame_text`
 * events); the Dart side ([SupyTextPatternMatcher]) runs the regex patterns
 * over it. This view therefore exposes only camera lifecycle: torch / pause /
 * resume + a `preview_started` event. Detection runs on the analyzer thread;
 * events are marshalled to main only at the sink boundary. Single ML Kit
 * client per PlatformView, closed in dispose().
 *
 * Arabic OCR is intentionally not supported on Android (ML Kit is Latin-only);
 * iOS covers `ar` via Vision. See `docs/MIGRATION.md`.
 */
class SupyDataCaptureScannerView(
    private val context: Context,
    viewId: Int,
    @Suppress("UNUSED_PARAMETER") creationParams: Map<String, Any?>?,
    messenger: BinaryMessenger,
    private val activityHolder: ActivityHolder,
) : PlatformView, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {

    private val container: FrameLayout = FrameLayout(context).apply {
        setBackgroundColor(Color.BLACK)
    }

    private val previewView: PreviewView = PreviewView(context).apply {
        scaleType = PreviewView.ScaleType.FILL_CENTER
        layoutParams = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.MATCH_PARENT,
        )
    }

    private val methodChannel: MethodChannel = MethodChannel(
        messenger,
        "$DATACAPTURE_CHANNEL_PREFIX/$viewId",
    ).also { it.setMethodCallHandler(this) }

    private val eventChannel: EventChannel = EventChannel(
        messenger,
        "$DATACAPTURE_CHANNEL_PREFIX/$viewId/events",
    ).also { it.setStreamHandler(this) }

    private val mainHandler: Handler = Handler(Looper.getMainLooper())

    // One ML Kit text-recognizer client per PlatformView. Closed in dispose().
    private val recognizer: TextRecognizer =
        TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)

    private var cameraController: LifecycleCameraController? = null
    private var eventSink: EventChannel.EventSink? = null
    private var previewStartedAnnounced: Boolean = false

    private val deviceTier: DeviceTier = DeviceTier.detect(context)
    private val analyzerExecutor = Executors.newSingleThreadExecutor { r ->
        Thread(r, "supy-datacapture-analyzer").apply { isDaemon = true }
    }
    // OCR is heavier than barcode decode, so cap the analyzer cadence on the
    // per-tier FPS ceiling (falls back to ~10fps when the tier is uncapped) to
    // keep the analyzer thread from queueing behind a slow recognition pass.
    private val frameIntervalNs: Long =
        1_000_000_000L / (deviceTier.analyzerFpsCap() ?: 10)
    private var lastAnalyzedNs: Long = 0L

    init {
        container.addView(previewView)
        startCamera()
    }

    // ───────── Camera lifecycle ─────────

    private fun startCamera() {
        val activity = activityHolder.activity
        if (activity == null) {
            emitError("camera_unavailable", "Host Activity is not attached yet")
            return
        }
        val lifecycleOwner = activity as? LifecycleOwner
        if (lifecycleOwner == null) {
            emitError("camera_unavailable", "Host Activity is not a LifecycleOwner")
            return
        }

        val controller = LifecycleCameraController(context).apply {
            setImageAnalysisBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
            setImageAnalysisAnalyzer(analyzerExecutor, AnalyzerImpl())
            bindToLifecycle(lifecycleOwner)
        }
        previewView.controller = controller
        cameraController = controller

        previewView.previewStreamState.observe(lifecycleOwner) { state ->
            if (state == PreviewView.StreamState.STREAMING && !previewStartedAnnounced) {
                previewStartedAnnounced = true
                emitPreviewStarted(controller)
            }
        }
    }

    private fun emitPreviewStarted(controller: LifecycleCameraController) {
        val flashAvailable = runCatching {
            controller.cameraInfo?.hasFlashUnit() == true
        }.getOrDefault(false)
        sendEvent(
            mapOf(
                "type" to "preview_started",
                "flashAvailable" to flashAvailable,
            ),
        )
    }

    private fun emitError(code: String, message: String) {
        sendEvent(
            mapOf(
                "type" to "error",
                "code" to code,
                "message" to message,
            ),
        )
    }

    private fun sendEvent(payload: Map<String, Any?>) {
        val sink = eventSink ?: return
        if (Looper.myLooper() == Looper.getMainLooper()) {
            sink.success(payload)
        } else {
            mainHandler.post { eventSink?.success(payload) }
        }
    }

    // ───────── Analyzer ─────────

    private inner class AnalyzerImpl : ImageAnalysis.Analyzer {

        @SuppressLint("UnsafeOptInUsageError")
        @OptIn(ExperimentalGetImage::class)
        @WorkerThread
        override fun analyze(image: ImageProxy) {
            val now = System.nanoTime()
            if (now - lastAnalyzedNs < frameIntervalNs) {
                image.close()
                return
            }
            lastAnalyzedNs = now

            val mediaImage = image.image
            if (mediaImage == null) {
                image.close()
                return
            }

            val rotation = image.imageInfo.rotationDegrees
            // ML Kit reports geometry in the upright (rotation-applied) frame,
            // so the dimensions to normalize against swap for 90/270.
            val imgWidth = if (rotation == 90 || rotation == 270) image.height else image.width
            val imgHeight = if (rotation == 90 || rotation == 270) image.width else image.height

            val input = InputImage.fromMediaImage(mediaImage, rotation)
            recognizer.process(input)
                .addOnSuccessListener { recognized ->
                    val tree = buildTree(recognized, imgWidth, imgHeight)
                    sendEvent(tree + ("type" to "frame_text"))
                }
                .addOnFailureListener { e ->
                    emitError("unknown", e.message ?: "Text recognition failed")
                }
                .addOnCompleteListener {
                    image.close()
                }
        }
    }

    /**
     * Builds the block → line → element tree consumed by
     * `SupyRecognizedText.fromMap`. Boxes normalized to `[0..1]`, top-left
     * origin. Elements are always included on the live path — the Dart matcher
     * only reads block/line/fullText, but shipping elements keeps the wire tree
     * identical to the standalone `recognizeText` shape.
     */
    private fun buildTree(text: Text, imgWidth: Int, imgHeight: Int): Map<String, Any?> {
        val blocks = text.textBlocks.map { block ->
            val lines = block.lines.map { line ->
                val elements = line.elements.map { element ->
                    mapOf(
                        "text" to element.text,
                        "boundingBox" to normRect(element.boundingBox, imgWidth, imgHeight),
                    )
                }
                mapOf(
                    "text" to line.text,
                    "boundingBox" to normRect(line.boundingBox, imgWidth, imgHeight),
                    "elements" to elements,
                )
            }
            mapOf(
                "text" to block.text,
                "boundingBox" to normRect(block.boundingBox, imgWidth, imgHeight),
                "lines" to lines,
            )
        }
        return mapOf("fullText" to text.text, "blocks" to blocks)
    }

    private fun normRect(rect: Rect?, imgWidth: Int, imgHeight: Int): Map<String, Double> {
        if (rect == null || imgWidth <= 0 || imgHeight <= 0) {
            return mapOf("left" to 0.0, "top" to 0.0, "width" to 0.0, "height" to 0.0)
        }
        val w = imgWidth.toDouble()
        val h = imgHeight.toDouble()
        return mapOf(
            "left" to rect.left / w,
            "top" to rect.top / h,
            "width" to rect.width() / w,
            "height" to rect.height() / h,
        )
    }

    // ───────── MethodChannel ─────────

    @MainThread
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "pause" -> {
                cameraController?.unbind()
                result.success(null)
            }
            "resume" -> {
                val lifecycleOwner = activityHolder.activity as? LifecycleOwner
                if (lifecycleOwner == null) {
                    result.error("camera_unavailable", "No LifecycleOwner", null)
                } else {
                    cameraController?.bindToLifecycle(lifecycleOwner)
                    result.success(null)
                }
            }
            "setTorch" -> {
                val on = call.argument<Boolean>("on") ?: false
                cameraController?.enableTorch(on)
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    // ───────── EventChannel ─────────

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        val controller = cameraController ?: return
        if (previewView.previewStreamState.value == PreviewView.StreamState.STREAMING &&
            !previewStartedAnnounced
        ) {
            previewStartedAnnounced = true
            emitPreviewStarted(controller)
        }
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    // ───────── PlatformView ─────────

    override fun getView(): View = container

    @MainThread
    override fun dispose() {
        methodChannel.setMethodCallHandler(null)
        runCatching { eventSink?.endOfStream() }
        eventSink = null
        eventChannel.setStreamHandler(null)
        cameraController?.clearImageAnalysisAnalyzer()
        cameraController?.unbind()
        previewView.controller = null
        cameraController = null
        runCatching { recognizer.close() }
        runCatching { analyzerExecutor.shutdown() }
        container.removeAllViews()
    }

    companion object {
        private const val DATACAPTURE_CHANNEL_PREFIX = "io.supy.scanner/v1/datacapture"
    }
}
